Java sprint 7 v1.0

1EVILGUN1 opened the pull request:
Созданы классы : FileBackedTaskManager, TaskConverter, TaskData. Класс enum  : TaskType. Классы исключения: ManagerSaveException, NotFoundException. Переделаны некоторые классы: InMemoryTaskManager,Manager, Task, Subtask, Epic

1EVILGUN1 committed:
Java sprint 7 v1.0

Созданы классы : FileBackedTaskManager, TaskConverter, TaskData. Класс enum  : TaskType. Классы исключения: ManagerSaveException, NotFoundException. Переделаны некоторые классы: InMemoryTaskManager,Manager, Task, Subtask, Epic

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/InMemoryTaskManager.java:
```diff
@@ -13,10 +14,18 @@
-    final Map<Integer, Epic> epics = new HashMap<>();
-    final Map<Integer, Subtask> subtasks = new HashMap<>();
-    final HistoryManager historyManager = Manager.getDefaultHistory();
+    final Map<Integer, Task> tasks ;
```
стоит все поля тут сделать protected что бы у наслденика был доступ
Так же удали лишний пробел перед  ;

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+public class FileBackedTaskManager extends InMemoryTaskManager {
+    HashMap<TaskType, TaskConverter> converters;
+    private final File file;
+    public FileBackedTaskManager() {
```
Достаточно 1 конструктора, что бы однозначно обьявить обьект
	public FileBackedTaskManager(File file) {
		this.file = file;
		 this(Manager.getDefaultHistory());
	}

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+import ru.yandex.javacource.golotin.schedule.model.*;
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+    HashMap<TaskType, TaskConverter> converters;
```
это лишнее

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+
+    public static FileBackedTaskManager loadFromFile(File file) {
+        FileBackedTaskManager manager = new FileBackedTaskManager(file);
+        manager.initialization();
```
вызывешь initialization() а потом в нем сразу  loadFromFile();
просто несколько методо проксируют и все. 
Делай сразу тут логику.

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/TaskData.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.List;
+
+public class TaskData {
```
Этот класс лишний

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+                    break;
+                }
+            }
+            String line = reader.readLine();// История
```
получается тоже лишнее, в здании больше не нужн овосстанавливать историю

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+                String line = reader.readLine();
+                final Task task = fromString(line);
+                final int id = task.getId();
+                if (task.getType() == TaskType.TASK) {
```
Ты восстанавливаешь только задачи, а как же сабтаски и Эпик?

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+
+    private Task fromString(String value) {
+        final String[] columns = value.split(",");
+        int id = 0;
```
поле не используется, а должно, ведь стоит восстанавливать задачи именно с тем id которая в истории

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        Task task = null;
+        switch (type) {
+            case TASK:
+                task = new Task(name, status, description);
```
расширь контракт и создавай сразу с id

YazykovMikhail replied:
Для всех задач

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        return sb;
+    }
+
+    static List<Integer> historyFromString(String value) {
```
метод не используется, стоит удалить

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        return task;
+    }
+
+    static String toString(HistoryManager manager) {
```
метод не используется, стоит удалить

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        counterId = maxId;// генератор
+    }
+
+    public static final String TASK_CSV = "task.csv";
```
поля обьявлюятся в начале класса, а констнаты самыми первыми, но эт ополе пока тут не нужно.

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/Manager.java:
```diff
@@ -2,7 +2,7 @@
 
 public class Manager {
     public static TaskManager getDefaults() {
```
Теперь работа поумолчанию это работа через FileBackedTaskManager
	public static TaskManager getDefault() {
		return new FileBackedTaskManager(new File("resources/task.csv"));
	}

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        loadFromFile();
+    }
+
+    public static FileBackedTaskManager loadFromFile(File file) {
```
У тебя не плохая реализация, но есть что улучшить, можешь разобрать код и приметь к себе:

```


	public static FileBackedTaskManager loadFromFile(File file) {
		final FileBackedTaskManager taskManager = new FileBackedTaskManager(file);
		try {
			final String csv = Files.readString(file.toPath());
			final String[] lines = csv.split(System.lineSeparator());
			int generatorId = 0;
			for (int i = 1; i < lines.length; i++) {
				String line = lines[i];
				if (line.isEmpty()) {
					break;
				}
				final Task task = taskFromString(line);
				final int id = task.getId();
				if (id > generatorId) {
					generatorId = id;
				}
				taskManager.addAnyTask(task);
			}
			for (Map.Entry<Integer, Subtask> e : taskManager.subtasks.entrySet()) {
				final Subtask subtask = e.getValue();
				final Epic epic = taskManager.epics.get(subtask.getEpicId());
				epic.addSubtaskId(subtask.getId());
			}
			taskManager.generatorId = generatorId;
		} catch (IOException e) {
			throw new ManagerSaveException("Can't read form file: " + file.getName(), e);
		}
		return taskManager;
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+
+    private void saveToFile() {// Сохранение в файл
```
Отлично, только ты забыл сохранять Заголовок и только задачи, а как же Эпик и Сабтаск:
```
protected void save() {
		try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
			writer.write(HEADER);
			writer.newLine();

			for (Map.Entry<Integer, Task> entry : tasks.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			for (Map.Entry<Integer, Subtask> entry : subtasks.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			for (Map.Entry<Integer, Epic> entry : epics.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			writer.newLine();
		} catch (IOException e) {
			throw new ManagerSaveException("Can't save to file: " + file.getName(), e);
		}
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+import ru.yandex.javacource.golotin.schedule.exception.ManagerSaveException;
+import ru.yandex.javacource.golotin.schedule.model.*;
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
```
private static final String HEADER = "id,type,name,status,description,epic";

YazykovMikhail replied:
Вынеси заголовок в константу

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        return manager;
+    }
+    @Override
+    public List<Task> getTasks() {
```
Все методы переопределять не нужно, только те где вызывается метод save
Ведь тут ничего не меняется

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public void updateTask(Task task) {
```
В методах update тоже нужно вызывать save ведь состояние меняется

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public void deleteTask(int id) {
```
В методах delete тоже нужно вызывать save ведь состояние меняется

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+    public void deleteSubtask(int id) {
+        super.deleteSubtask(id);
+    }
+
```
Просто метод реализации toString:
```
public static String toString(Task task) {
		return task.getId() + "," + task.getType() + "," + task.getName() + "," + task.getStatus() + "," + task.getDescription() + "," + (task.getType().equals(TaskType.SUBTASK) ? task.getEpicId() : "");
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,169 @@
+        return STR."\{task.getId()},\{task.getName()},\{task.getDescription()},\{task.getEpicId()}";
+    }
+
+    private Task fromString(String value) {
```
Получается будет вот так:
```

	public static Task taskFromString(String value) {
		final String[] values = value.split(",");
		final int id = Integer.parseInt(values[0]);
		final TaskType type = TaskType.valueOf(values[1]);
		final String name = values[2];
		final TaskStatus status = TaskStatus.valueOf(values[3]);
		final String description = values[4];
		if (type == TaskType.TASK) {
			return new Task(id, name, description, status);
		}
		if (type == TaskType.SUBTASK) {
			final int epicId = Integer.parseInt(values[5]);
			return new Subtask(id, name, description, status, epicId);
		}

		return new Epic(id, name, description, status);
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/converter/TaskConverter.java:
```diff
@@ -0,0 +1,14 @@
+import ru.yandex.javacource.golotin.schedule.model.Task;
+import ru.yandex.javacource.golotin.schedule.model.TaskType;
+
+public class TaskConverter {
```
метод  getType() тут не нужен, а ради одного метода можн оне делать утилитный класс, показал как его можно сделать приватным методом

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/exception/ManagerSaveException.java:
```diff
@@ -0,0 +1,7 @@
+    public ManagerSaveException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}
```
Стоит добавить конструкторы, базовые:
```

	public ManagerSaveException() {
	}

	public ManagerSaveException(String message) {
		super(message);
	}

	public ManagerSaveException(String message, Throwable cause) {
		super(message, cause);
	}

	public ManagerSaveException(Throwable cause) {
		super(cause);
	}
```

1EVILGUN1 committed:
Java sprint 7 v1.1

Удалены классы: TaskConverter, TaskData.
Обновлены классы: FileBackedTaskManager,Manager.
Добавлены методы в классы исключения; ManagerSaveException, NotFoundException
Добавлена перегрузка конструкторов: Task,Subtask,Epic

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/Subtask.java:
```diff
@@ -9,19 +9,25 @@ public Subtask(String name, Status status, String description,int epicId) {
         setEpicId(epicId);
     }
-
+    public Subtask(int id,String name,String description,Status status,Integer epicId){
```
public Subtask(int id, String name, String description, Status status, Integer epicId) { - обрати внимание на пробелы, они нужны по кодстайлу

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,211 @@
+        }
+    }
+
+    private void loadFromFile() {// Чтение из в файла
```
У тебя два метода 
с названием loadFromFile
В задании сигнатура метода выглядит так:
loadFromFile(File file) 
Оставь однин верный метод

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,211 @@
+
+            for (Map.Entry<Integer, Task> entry : tasks.entrySet()) {
+                final Task task = entry.getValue();
+                writer.write(CSVTaskFormat.toString(task));
```
А ты запускал код? тестировал? У тебя же нет класса CSVTaskFormat, код не запустится. 
Или сделать этот класс и вынести туда статический метод превращения в строку задач. 
Или просто тут сделать вот так:
toString(task)
Ведь метод toString у теб написан прям тут рядом приватным методом.

YazykovMikhail replied:
Поправь для всех

YazykovMikhail commented:
Одно из заданий "Проверяем работу нового менеджера" стоит его выполнять, сейчас не работает. Если есть непонимнаие по ревью всегда можно спросить у ревьюера или если нужна помощь с заданием попросить наставника

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/Epic.java:
```diff
@@ -23,15 +27,15 @@ public void removeSubtask(int id) {
-        return "Epic{" +
-                "subtaskIds=" + subtaskIds +
-                '}';
+        return STR."Epic{subtaskIds=\{subtaskIds}\{'}'}";
```
а что такое 
STR.
?
У меня тут ошибка ситаксиса, как и дрвугих местах где такая же конструкция

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/Epic.java:
```diff
@@ -23,15 +27,15 @@ public void removeSubtask(int id) {
         return subtaskIds;
     }
 
+    public TaskType getType() {return TaskType.EPIC;}
```
Кодстайл стоит соблюдать:
```
  public TaskType getType() {
        return TaskType.EPIC;
    }
```

YazykovMikhail replied:
Поправь во всех задачах

YazykovMikhail commented:
Стоит лобавить папку 
resources
что бы она добавилась нужно в ней создать файл task.csv закомитать и запушить.

YazykovMikhail commented:
Раз я сразу не сказал, то сейчас не буду просить поправить тесты, но в следующем задании стоит поправить, что бы тесты работали, если не получится самостоятельно, запроси помощи у наставника

1EVILGUN1 committed:
Java sprint 7 v1.2

Обновлен класс FileBackedTaskManager
Убран String Templates и переделан под Java 11
reformat code классов Epic,Subtask,Task, FileBackedTaskManager

1EVILGUN1 committed:
Java sprint 7 v1.2.1

переделан путь папки resources. и путь в классе manager для сохранения файла task.csv
